dockerfiles/csharp/Dockerfile

```diff
@@ -18,7 +18,7 @@ RUN apt-get update \
   && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF \
   && apt-get update && apt-get install --no-install-recommends -y apt-transport-https=1.2.26 \
   # OmniSharp runs on Mono and requires the .NET SDK 2.0.0.
-  && apt-get update && apt-get install --no-install-recommends -y mono-complete=5.12.0.226-0xamarin3+ubuntu1604b1 dotnet-sdk-2.0.0=2.0.0-1 nodejs=10.2.1-1nodesource1 \
+  && apt-get update && apt-get install --no-install-recommends -y mono-complete=5.12.0.226-0xamarin3+ubuntu1604b1 dotnet-sdk-2.0.0=2.0.0-1 nodejs=10.3.0-1nodesource1 \
```

Comment:
Since the nodesource apt repos delete old minor and patch versions, we should probably not pin this or download the binary manually like we do for the buildkite agent

Reply:
Sure, but I believe this is true for most package repositories - they only keep the latest version around. (i.e. why don't we also un-pin other packages like `git`), etc?

Reply:
Are you sure that is the case for git?

Reply:
I think most repositories behave this way: https://github.com/influxdata/influxdb/issues/9229#issuecomment-352576609. If you want to pin a specific version, I think you need download the `deb` directly from somewhere, or use an index that explicitly keeps old versions around.

Reply:
I also don't think there is much risk of tools like `git` or `curl` making breaking changes to their api that often...

Reply:
And if they did, the distro maintainer will mostly likely not upgrade the version of the package in the index (i.e. you need to upgrade the distro itself to get the new versions)

Reply:
My concern is less about git or curl making breaking changes but more about building countless new image versions with new hashes on every commit that are not cached.

For nodesource it was mentioned that this is a limitation of the `reprepro` tool they use to host the repository and using a different tool like `aptly` would solve it. If the examples you linked is also a custom repository they might have the same situation, but most packages we use are in the official repositories so this doesn't necessarily apply to them - while it may be popular for custom repositories, it's not necessarily true for the official ones (otherwise why would apt-get support version pinning at all?).